R/nm-tables.R

```diff
@@ -0,0 +1,131 @@
+
+#' Read all tables and input data
+#'
+#' Reads in the input data set and all table output files from a NONMEM run.
+#' @return A named list of tibbles. The first element will always be named
+#'   `data` and will contain the input data set. Subsequent elements will be
+#'   named for the file from which they were loaded, with `get_model_id(.mod)`
+#'   (and `.`) removed from the beginning and end, if present. For example, a
+#'   model named `001.ctl` that generated a table named `001.tab` will have the
+#'   relevant element named `tab`.
+#' @param .mod Either a `bbi_nonmem_model` or `bbi_nonmem_summary` object
+#' @param .files Character vector of file paths to table files to read in.
+#'   Defaults to calling [nm_table_files()] on `.mod`, which will parse all file
+#'   names from `$TABLE` blocks in the control stream.
+#' @importFrom purrr compact map_chr
+#' @importFrom stringr str_replace
+#' @export
+nm_tables <- function(
+  .mod,
+  .files = nm_table_files(.mod)
+) {
+  check_model_object(.mod, c(NM_MOD_CLASS, NM_SUM_CLASS))
+  checkmate::assert_character(.files)
+
+  # make paths absolute
+  .files <- map_chr(.files, ~{
+    if (fs::is_absolute_path(.x)) {
+      return(.x)
+    } else {
+      return(file.path(get_output_dir(.mod), .x))
+    }
+  })
+
+  # read in input data
+  res <- list(
+    data = nm_data(.mod)
+  )
+
+  # build names for table elements
+  .n <- .files %>%
+    basename() %>%
+    str_replace(glue("^{get_model_id(.mod)}"), "") %>%
+    str_replace(glue("{get_model_id(.mod)}$"), "") %>%
+    make.names() %>%
+    str_replace(glue("^\\Q.\\E"), "") %>%
+    str_replace(glue("\\Q.\\E$"), "")
+
+  # read in each table file
+  for (.i in 1:length(.files)) {
+    res[[.n[.i]]] <- nm_file(.files[.i], .est_method = "fail")
+  }
+  return(compact(res))
+}
+
+#' @describeIn nm_tables Extract paths to table output files from NONMEM control
+#'   stream, and optionally check if the files exist.
+#' @param .check_exists If `TRUE` check if files exist.
+#' @importFrom stringr str_extract str_replace
+#' @importFrom purrr map_chr
+#' @export
+nm_table_files <- function(.mod, .check_exists = TRUE) {
+  .p <- get_model_path(.mod)
+  .l <- parse_ctl_to_list(.p)
+
+  # get file names from table statements and construct paths
+  .f <- .l[names(.l) == "TABLE"] %>%
+    map_chr(~paste(.x, collapse = " ")) %>%
+    str_extract("\\bFILE\\s*=\\s*([^ ]+)") %>%
+    str_replace("\\bFILE\\s*=\\s*", "") %>%
+    str_replace("^\\Q./\\E", "") %>%
+    file.path(get_output_dir(.mod, .check_exists = .check_exists), .)
```

Comment:
Similar subjective comment as above: I'd find it more readable to have  the `get_output_dir` call out of the loop since it doesn't change.